src/tx_manager.rs

```diff
@@ -0,0 +1,704 @@
+//! stages, thereby saving gas. Most often the reason for the failed transaction
+//! is the lack of gas. Unfortunately, this algorithm isn't too user-friendly
+//! because users of contracts with caching must track cached
+//! [`TransactionId`]s, and increment their number on each successful execution.
+//! [`TransactionManager`] is intended to help solve this problem.
+//!
+//! In simple words, [`TransactionManager`] stores pairs of [`msg::source()`] &
+//! some data that a contract developer wants to save for this
+//! [`msg::source()`]. The manager has a limit at which it starts to replace old
+//! pairs with new ones. While doing all this, it tracks transaction identifiers
+//! (or more specifically slices of identifiers) for all pairs. Thus, having
+//! only [`msg::source()`], it's possible to acquire [`TransactionGuard`] with
+//! [`Stepper`] inside, and by calling [`Stepper::step()`] always get the
+//! correct [`TransactionId`] regardless of whether the transaction is new or
+//! cached.
+//!
+//! [`msg::source()`]: gstd::msg::source
+
+use ahash::AHasher;
+use core::num::{NonZeroU32, NonZeroUsize};
+use gstd::{hash::BuildHasherDefault, prelude::*, ActorId};
+use indexmap::{map::MutableKeys, IndexMap};
+use offset::Offset;
+
+mod offset;
+
+/// The default transaction limit.
+///
+/// The contract storage is limited, so cached transactions can't take up all
+/// the space. Note that although this limit may be enough for most example
+/// cases, the real contract structure may still be very different, so it's
+/// recommended to calculate an individual limit for each contract, and set it
+/// with [`TransactionManager::new_with_custom_limit()`].
+pub const DEFAULT_TX_LIMIT: NonZeroU32 = unsafe { NonZeroU32::new_unchecked(2u32.pow(16)) };
+/// The maximum transaction limit.
+///
+/// A transaction limit mustn't be more than [`u32::MAX`]` / `[`u8::MAX`]` - 1`.
+/// With the current contract memory limit (32 MB), it's impossible to store
+/// even the half of this amount, so consider this just as an additional
+/// restriction.
+///
+/// The reason for it is the [`TransactionManager`]'s logic for
+/// [`TransactionId`] traversing. The manager divides available
+/// [`TransactionId`]s by [`u8::MAX`], saves only division indexes, and
+/// multiplies them by [`u8::MAX`] to get actual [`TransactionId`]s. Hence to
+/// avoid the [`u32`] overflow, the maximum amount of cached transactions
+/// multiplied by [`u8::MAX`] mustn't be more than
+/// [`u32::MAX`]` / `[`u8::MAX`]` - 1`.
+pub const MAX_TX_LIMIT: NonZeroU32 = unsafe { NonZeroU32::new_unchecked(Offset::MAX) };
+
+/// A transaction identifier.
+// TODO: change to `u32` after the SFT refactor.
+pub type TransactionId = u64;
+
+impl<T> Default for TransactionManager<T> {
+    fn default() -> Self {
+        Self {
+            actors_for_tx: IndexMap::default(),
+            tx_limit: DEFAULT_TX_LIMIT.try_into().unwrap(),
```

Comment:
Is this conversion not fallible? Can we replace it with `into()`?

Reply:
It should be replaced with `.into()` by the compiler for the `wasm32-unknown-unknown` target. But since tests are compiled for the 64-bit host, `.try_into().unwrap()` is necessary here.